sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -183,34 +182,111 @@ private[sql] object ArrowConverters {
+   */
   private[sql] def toDataFrame(
-      payloadRDD: JavaRDD[Array[Byte]],
+      arrowBatchRDD: JavaRDD[Array[Byte]],
       schemaString: String,
       sqlContext: SQLContext): DataFrame = {
-    val rdd = payloadRDD.rdd.mapPartitions { iter =>
+    val schema = DataType.fromJson(schemaString).asInstanceOf[StructType]
+    val timeZoneId = sqlContext.sessionState.conf.sessionLocalTimeZone
+    val rdd = arrowBatchRDD.rdd.mapPartitions { iter =>
       val context = TaskContext.get()
-      ArrowConverters.fromPayloadIterator(iter.map(new ArrowPayload(_)), context)
+      ArrowConverters.fromBatchIterator(iter, schema, timeZoneId, context)
     }
-    val schema = DataType.fromJson(schemaString).asInstanceOf[StructType]
     sqlContext.internalCreateDataFrame(rdd, schema)
   }
+
+  /**
+   * Read a file as an Arrow stream and parallelize as an RDD of serialized ArrowRecordBatches.
+   */
+  private[sql] def readArrowStreamFromFile(
+      sqlContext: SQLContext,
+      filename: String): JavaRDD[Array[Byte]] = {
+    val fileStream = new FileInputStream(filename)
+    try {
+      // Create array so that we can safely close the file
+      val batches = getBatchesFromStream(fileStream.getChannel).toArray
+      // Parallelize the record batches to create an RDD
+      JavaRDD.fromRDD(sqlContext.sparkContext.parallelize(batches, batches.length))
+    } finally {
+      fileStream.close()
+    }
+  }
+
+  /**
+   * Read an Arrow stream input and return an iterator of serialized ArrowRecordBatches.
+   */
+  private[sql] def getBatchesFromStream(in: SeekableByteChannel): Iterator[Array[Byte]] = {
+
+    // TODO: this could be moved to Arrow
+    def readMessageLength(in: ReadChannel): Int = {
+      val buffer = ByteBuffer.allocate(4)
+      if (in.readFully(buffer) != 4) {
+        return 0
+      }
+      MessageSerializer.bytesToInt(buffer.array())
+    }
+
+    // TODO: this could be moved to Arrow
+    def loadMessage(in: ReadChannel, messageLength: Int, buffer: ByteBuffer): Message = {
+      if (in.readFully(buffer) != messageLength) {
+        throw new java.io.IOException(
+          "Unexpected end of stream trying to read message.")
+      }
+      buffer.rewind()
+      Message.getRootAsMessage(buffer)
+    }
+
+
+    // Create an iterator to get each serialized ArrowRecordBatch from a stream
+    new Iterator[Array[Byte]] {
+      val inputChannel = new ReadChannel(in)
+      var batch: Array[Byte] = readNextBatch()
+
+      override def hasNext: Boolean = batch != null
+
+      override def next(): Array[Byte] = {
+        val prevBatch = batch
+        batch = readNextBatch()
+        prevBatch
+      }
+
+      def readNextBatch(): Array[Byte] = {
+        val messageLength = readMessageLength(inputChannel)
+        if (messageLength == 0) {
+          return null
+        }
+
+        val buffer = ByteBuffer.allocate(messageLength)
+        val msg = loadMessage(inputChannel, messageLength, buffer)
```

Comment:
The only issue with this is that it is aware that a message is preceded by the message length and that a length of zero indicates no more messages.  Ideally, this logic would be abstracted to Arrow...

Reply:
Can we use `MessageChannelReader.readNextMessage()`?

Reply:
We can't really use it as is for a couple of reasons.  It just returns a flatbuffer `Message`, which means that to get it back into another buffer, it requires doing a lot of work with a `FlatBufferBuilder` and some special low level logic that we don't want here.  Arrow has a method `serialize(WriteChannel out, Schema schema)` but then we would have to deserialize only to serialize again.  So it's best to just use the original buffer containing the message.

Reply:
I see.

IMHO the class still touches too many Arrow stream low-level details, but maybe necessary because of the performance improvement.

@BryanCutler does all the speed up of `createDataFrame` comes from avoiding extra copy here?

Reply:
I'll propose something different under https://github.com/apache/arrow/pull/2139, we can continue to discuss there